### PR TITLE
Calculate SpotPrice with Ratios and SigFigs

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #1155 

## Description

* Calculates SpotPrice with ratios to prevent trailing digits from giving incorrect results
* Utilizes constant in x/gamm/types to determine how many significant figures we calculate SpotPrice to (other functions can also use this SigFigs constant now)
* Added a test to show a circumstance in which the 9th digit is greater than 5, and with 8 sigfigs set should remove the 9th digit however still round the 8th digit up
* Added comment to explain why this change was added

### CODE REVIEW

**Comment by czarcas7ic** [comment]
Should I modify all the tests that failed to utilize the 8 sigfigs now?

**Comment by czarcas7ic** [comment]
I basically would just need to change this test here 
https://github.com/osmosis-labs/osmosis/blob/main/x/pool-incentives/keeper/keeper_test.go#L96 
and here 
https://github.com/osmosis-labs/osmosis/blob/main/x/gamm/keeper/keeper_test.go#L95

Just lmk what you would like and I will change accordingly

File: x/gamm/types/constants.go [hunk 1]
```diff
@@ -9,6 +9,8 @@ const (
 	MaxPoolAssets = 8
 
 	OneShareExponent = 18
+
+	SigFigsExponent = 8
```
Comment (alexanderbez) [inline x/gamm/types/constants.go:13]: Can we add a comment explaining what this variable is and what it's used for?

Comment (czarcas7ic) [reply x/gamm/types/constants.go:14]: Done!

File: x/gamm/pool-models/balancer/amm.go [hunk 2]
```diff
@@ -118,10 +122,10 @@ func (p Pool) SpotPrice(ctx sdk.Context, baseAsset, quoteAsset string) (sdk.Dec,
 		return sdk.Dec{}, errors.New("pool is misconfigured, got 0 weight")
 	}
 
-	numerator := base.Token.Amount.ToDec().Quo(base.Weight.ToDec())
-	denom := quote.Token.Amount.ToDec().Quo(quote.Weight.ToDec())
-	ratio := numerator.Quo(denom)
-
+	weightRatio := base.Weight.ToDec().Quo(quote.Weight.ToDec())
+	supplyRatio := base.Token.Amount.ToDec().Quo(quote.Token.Amount.ToDec())
+	fullRatio := supplyRatio.Quo(weightRatio)
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/amm.go:127]:
```suggestion
        // spot_price = (Base_supply / Weight_base) / (Quote_supply / Weight_quote)
        // spot_price = (weight_quote / weight_base) * (base_supply / quote_supply)
	invWeightRatio := quote.Weight.ToDec().Quo(base.Weight.ToDec())
	supplyRatio := base.Token.Amount.ToDec().Quo(quote.Token.Amount.ToDec())
	fullRatio := supplyRatio.Mul(invWeightRatio)
```

**Comment by czarcas7ic** [comment]
Test modified to include sigfigs, now pass

File: x/gamm/pool-models/balancer/amm.go [hunk 3]
```diff
@@ -118,10 +122,12 @@ func (p Pool) SpotPrice(ctx sdk.Context, baseAsset, quoteAsset string) (sdk.Dec,
 		return sdk.Dec{}, errors.New("pool is misconfigured, got 0 weight")
 	}
 
-	numerator := base.Token.Amount.ToDec().Quo(base.Weight.ToDec())
-	denom := quote.Token.Amount.ToDec().Quo(quote.Weight.ToDec())
-	ratio := numerator.Quo(denom)
-
+	// spot_price = (Base_supply / Weight_base) / (Quote_supply / Weight_quote)
+	// spot_price = (weight_quote / weight_base) * (base_supply / quote_supply)
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm.go:126]: They are equivalent! Its intended to show that they're the same, if you simplify the equation in that way

[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm.go:126]: The first equation is how its commonly quoted, but we're computing according to the second equation for the speed and improved precision due to the weight common denominator

**Review by ValarDragon — APPROVED** [review]
LGTM

**Pull request merged**